File: animals/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;
import java.util.Objects;

public class Animal implements WorldElement{
    private MapDirection orientation;
    private Vector2d position;
    public Animal() {
        this(new Vector2d(2,2));
    }
    public Animal(Vector2d position){
        this.orientation = MapDirection.NORTH;
        this.position = position;
    }
    public MapDirection getOrientation() {
        return orientation;
    }
    @Override
    public Vector2d getPosition() {
        return position;
    }

    public boolean isAt(Vector2d position){
        return (position.equals(this.position));
    }

    public void move(MoveDirection direction, MoveValidator validator){

        orientation = switch (direction) {
            case LEFT -> orientation.previous();
            case RIGHT -> orientation.next();
            case FORWARD,BACKWARD -> orientation;
        };

        Vector2d newPosition = switch (direction) {
            case FORWARD -> position.add(orientation.toUnitVector());
            case BACKWARD -> position.subtract(orientation.toUnitVector());
            case LEFT,RIGHT -> position;
        };

        if(validator.canMoveTo(newPosition)) {
            position = newPosition;
        }

    }
    @Override
    public String toString() {
        return switch (this.orientation){
            case NORTH  ->  "N";
            case SOUTH  ->  "S";
            case EAST   ->  "E";
            case WEST   ->  "W";
        };
    }

    @Override
    public String getFileName() {
        return switch (this.orientation){
            case NORTH  ->  "up.png";
            case SOUTH  ->  "down.png";
            case EAST   ->  "right.png";
            case WEST   ->  "left.png";
        };
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Animal animal = (Animal) o;
        return orientation == animal.orientation && Objects.equals(position, animal.position);
    }

    @Override
    public int hashCode() {
        return Objects.hash(orientation, position);
    }

}

File: animals_but_it_is_kotlin/src/main/kotlin/agh/ics/oop/model/Animal.kt
package agh.ics.oop.model

data class Animal(var position: Vector2d = Vector2d(2, 2)) : WorldElement{
    var orientation = MapDirection.NORTH

    fun isAt(position: Vector2d): Boolean = position == this.position

    fun move(direction: MoveDirection, validator: MoveValidator) {

        val orientationTest: MapDirection = when (direction) {
            MoveDirection.LEFT -> orientation.previous()
            MoveDirection.RIGHT -> orientation.next()
            MoveDirection.FORWARD, MoveDirection.BACKWARD -> orientation
        }
        val positionTest: Vector2d = when (direction) {  // pomijam logike z orginalnej czesci poruszania sie
            MoveDirection.FORWARD -> position + orientation.toUnitVector()
            MoveDirection.BACKWARD -> position - orientation.toUnitVector()
            MoveDirection.LEFT, MoveDirection.RIGHT -> position
        }

        if( validator.canMoveTo(positionTest) ){
            position = positionTest
            orientation = orientationTest
        }
        else{  // nastepuje odbicie
            validator.place(this)
        }

    }

    override fun toString(): String {
        return when (orientation) {
            MapDirection.NORTH -> "N"
            MapDirection.SOUTH -> "S"
            MapDirection.EAST -> "E"
            MapDirection.WEST -> "W"
        }
    }
}